test/test_topology.ml

```diff
@@ -0,0 +1,215 @@
+             List.map
+               (fun node -> NUMA.distance host (NUMA.node_of_cpu host c) node)
+               nodes
+           in
+           let worst = List.fold_left max 0 distances in
+           let best = List.fold_left min max_int distances in
+           let average = float (List.fold_left ( + ) 0 distances) /. float n in
+           {worst; best; nodes= []; average})
+    |> sum_costs
+  in
+  D.debug "Costs: %s" (Fmt.to_to_string pp costs) ;
+  let cpus = float @@ CPUSet.cardinal cpuset in
+  let nodes =
+    all_vms |> List.map (fun ((_, nodes), _) -> nodes) |> List.flatten
+  in
+  {costs with average= costs.average /. cpus; nodes}
+
+let cost_not_worse ~default c =
+  let worst = max default.worst c.worst in
+  let best = min default.best c.best in
+  let average = min default.average c.average in
+  D.debug "Default access times: %s; New plan: %s"
+    (Fmt.to_to_string pp default)
+    (Fmt.to_to_string pp c) ;
+  Alcotest.(
+    check int "The worst-case access time should not be changed from default"
+      default.worst worst) ;
+  Alcotest.(check int "Best case access time should not change" best c.best) ;
+  Alcotest.(
+    check (float 1e-3) "Average access times could improve" average c.average) ;
+  if c.best < default.best then
+    D.debug "The new plan has improved the best-case access time!" ;
+  if c.worst < default.worst then
+    D.debug "The new plan has improved the worst-case access time!" ;
+  if c.average < default.average then
+    D.debug "The new plan has improved the average access time!"
+
+let balancing nodes ~vms =
+  (* We expect to use many NUMA nodes when we have more VMs,
+   * more elaborate checks could be done on how well balanced the VMs are across NUMA nodes.
+   * Though in this case the default would always be best since it uses all nodes.
+   * *)
+  let nodes_used = List.sort_uniq compare nodes |> List.length in
+  min vms nodes_used |> float
+
+let check_aggregate_costs_not_worse (default, next, plans) ~cores ~vms =
+  let default = sum_costs default in
+  let next = sum_costs next in
+  cost_not_worse ~default next ;
+  let balancing_default = balancing default.nodes ~vms in
+  let balancing_next = balancing next.nodes ~vms in
+  let balancing_best = max balancing_next balancing_default in
+  Alcotest.(
+    check (float 1e-3) "Balancing could improve" balancing_best balancing_next) ;
+  if balancing_next > balancing_default then D.debug "Balancing has improved!" ;
+  let used_cpus =
+    plans |> List.map snd
+    |> List.fold_left CPUSet.union CPUSet.empty
+    |> CPUSet.cardinal
+  in
+  Alcotest.(check int "All vCPUs are used " cores used_cpus)
+
+let default_mem = Int64.shift_left 1L 30
+
+(* higher than default_mem *)
+let mem3 = Int64.div (Int64.mul 4L (Int64.shift_left 1L 34)) 3L
+
+let test_allocate ?(mem = default_mem) (expected_cores, h) ~vms () =
+  let memsize = Int64.shift_left 1L 34 in
+  let nodea = NUMA.nodes h |> Seq.map (fun _ -> memsize) |> Array.of_seq in
```

Comment:
Is there a need to create the node array?

I'd rather use `~memory:memsize` in L135 if there's no need.